Add diff inputs command

StephenDiligent opened the pull request:
This change set addresses issue [Add command to diff the inputs of 2 builds](https://github.com/simplesurance/baur/issues/120).

Of note:
* I introduced a new `func doInitDb(t *testing.T)` function in _diff_inputs_test.go_ that is a copy of `func runInitDb(t *testing.T)` in _command_test.go_. I did this because the `runInitDb` function is not accessible to _diff_inputs_diff.go_ and when I tried to refactor `runInitDb` into _util_test.go_ the linter complained that the function is not used. I believe that these issues are caused by the `// +build dbtest` comments at the start of the test files.
* I added a `t.SkipNow()` call to the `assertExitCode(t *testing.T, expected int)` function in _diff_inputs_test.go_ [here](https://github.com/simplesurance/baur/pull/204/files#diff-85b512ec00613009498c78f3ca8f2024R34). The reason for this was to stop the command from executing further when `exitFunc()` is called but also allow the test to complete. Using `os.Exit()` instead causes the test to exit without completing. Can this be solved in a better way?
* The Circle CI tests are currently failing due to the digest length check mentioned in the comments

fho commented:
I answere in https://github.com/simplesurance/baur/issues/120#issuecomment-703653693 regarding the output format.

StephenDiligent commented:
@fho I have updated the implementation to meet the updated test cases.
I have not implemented writing the input differences out yet but the inputs are being calculated and the correct exit codes should be returned from the execution.
The tests are failing in Circle CI with _ERROR: relation "migrations" already exists (SQLSTATE 42P07)_. I'm a bit confused because I am using `r := repotest.CreateBaurRepository(t, repotest.WithNewDB())` and the tests pass locally.
Also, I'm seeing intermittent failures locally with this error _ERROR: calculating digest for "app_one/.app.toml" failed: hash length is 95, expected length 96_. There could potentially be a bug in the digest calculation code?
Once again, thank you very much for your time and help!

StephenDiligent commented:
Update on this error _ERROR: calculating digest for "app_one/.app.toml" failed: hash length is 95, expected length 96_:
I have managed to reproduce this error here https://play.golang.org/p/9NL1doFpcRz
Sometimes the digests are calculated with a length of 95 characters and sometimes with 96 characters.
When the length check [here](https://github.com/simplesurance/baur/blob/master/internal/digest/digest.go#L63) is removed the digests can be converted back from a string correctly. Perhaps this check needs to be loosened to allow lengths of 95 and 96?
Given the digests are integral to how baur operates I'd be very hesitant to change how the digests are calculated.

fho commented:
> @fho I have updated the implementation to meet the updated test cases.
> The tests are failing in Circle CI with _ERROR: relation "migrations" already exists (SQLSTATE 42P07)_. I'm a bit confused because I am using `r := repotest.CreateBaurRepository(t, repotest.WithNewDB())` and the tests pass locally.
> Also, I'm seeing intermittent failures locally with this error _ERROR: calculating digest for "app_one/.app.toml" failed: hash length is 95, expected length 96_. There could potentially be a bug in the digest calculation code?
> Once again, thank you very much for your time and help!

I was able to reproduce the error also locally by starting a local postgresql container (`scripts/start-psql-container.sh`) and running `make dbtest`.
The error should be fixed with the commit that I pushed to your branch (https://github.com/simplesurance/baur/pull/204/commits/c3c2d4535c5578351e8c0eb0db8c2fac9fc9af5e).

The CI tests runs were successful on my commit.
Somehow the check status in the PR still shows the results of the CI runs from your previous commit though. I don't know why, seems to be an issue with github or the status that circleci submits.

fho commented:
> Update on this error _ERROR: calculating digest for "app_one/.app.toml" failed: hash length is 95, expected length 96_:
> I have managed to reproduce this error here https://play.golang.org/p/9NL1doFpcRz
> Sometimes the digests are calculated with a length of 95 characters and sometimes with 96 characters.
> When the length check [here](https://github.com/simplesurance/baur/blob/master/internal/digest/digest.go#L63) is removed the digests can be converted back from a string correctly. Perhaps this check needs to be loosened to allow lengths of 95 and 96?
> Given the digests are integral to how baur operates I'd be very hesitant to change how the digests are calculated.

I'll check that tomorrow, probably the length check is wrong.

StephenDiligent commented:
@fho Thanks for fixing the db issue 🙂. This is now ready for review.

fho commented:
> > Update on this error _ERROR: calculating digest for "app_one/.app.toml" failed: hash length is 95, expected length 96_:
> > I have managed to reproduce this error here https://play.golang.org/p/9NL1doFpcRz
> > Sometimes the digests are calculated with a length of 95 characters and sometimes with 96 characters.
> > When the length check [here](https://github.com/simplesurance/baur/blob/master/internal/digest/digest.go#L63) is removed the digests can be converted back from a string correctly. Perhaps this check needs to be loosened to allow lengths of 95 and 96?
> > Given the digests are integral to how baur operates I'd be very hesitant to change how the digests are calculated.
> 
> I'll check that tomorrow, probably the length check is wrong.

@StephenDiligent 
It was actually a bug, leading zeroes were truncated from the hash accidentally. I fixed it here: https://github.com/simplesurance/baur/pull/208 
Thanks a lot for finding that issue :-)

fho commented on internal/command/diff_inputs.go:
```diff
@@ -0,0 +1,390 @@
+	return "", "", s
+}
+
+func (c *diffInputsCmd) getTaskRunInputs(app, task, runID string) (*baur.Inputs, *storage.TaskRunWithID, error) {
```
Currently `getTaskRunInputs()` is called 2x when the command is invoked.
Both times `mustFindRepository()` is called, `mustArgToTask()` can be called twice depending on the commandline parameters.
This means everytime `getTaskRunInputs()` is called, the repository has to be located in the filesystem, the baur config has to be loaded and passed, the application config files have to be found, they have to be read and parsed.

Doing these operations multiple times is a lot of unnecessary overhead which can be prevented:

- call `mustFindRepository()` only 1x in the parent function
- if both commandline parameters refer to tasks in the repository (`<APPNAME>.<TASK-NAME>`), call the `mustArgToTask()` function only 1x and pass both parameters (`mustArgToTask(param1, param2)`)
- would be good to call `mustNewCompatibleStorage(repo)` also only 1x in the parent function, does not have such a  huge overhead but it's also unnecessary to create the psql client 2x

fho commented on internal/command/diff_inputs.go:
```diff
@@ -0,0 +1,390 @@
+	return taskRun, nil
+}
+
+func getSorters() []*storage.Sorter {
```
this functions seems like it could be replace with a variable, it always returns a slice with the same content

```go
  var defSorters = []*storage.Sorter{
          storage.Sorter{
                  Field: storage.FieldID,
                  Order: storage.OrderDesc,
          },
  }
```

fho commented on internal/command/diff_inputs.go:
```diff
@@ -0,0 +1,390 @@
+
+type differenceType int
+
+const (
+	digestMismatch differenceType = iota
+	existsInOneOnly
+	existsInTwoOnly
+)
```
The first List of Inputs is the base of comparison, therefore we could name the difference types like:

```suggestion
const (
	digestMismatch differenceType = iota
	removed
	added
)
```

fho reviewed (changes requested):
@StephenDiligent great job, the new command will be very useful to figure out which input changes triggered task runs.
The test coverage is awesome. I think it is now the command with the best coverage. :-)
I also tried the command with the baur-example repository and everything works :tada:.

I added some comments about possible improvements.

StephenDiligent commented on internal/command/diff_inputs.go:
```diff
@@ -0,0 +1,390 @@
+// parseDiffSpec splits an argument into the app, task and runID components.
+// It relies on the fact that the arguments have already been validated
+// in the cobra.Command
+func (c *diffInputsCmd) parseDiffSpec(s string) (app, task, runID string) {
```
Sounds good, I'd prefer if this was deferred

StephenDiligent commented:
@fho thank you for your feedback, I have addressed all of your comments.
It was really handy having the tests when performing large refactorings like these.

fho commented:
@StephenDiligent I'll check it out beginning of next week.

fho commented on internal/command/diff_inputs.go:
```diff
@@ -0,0 +1,387 @@
+		}
+
+		inputs = baur.NewInputs(baur.InputAddStrIfNotEmpty(inputFiles, c.inputStr))
+	} else {
```
instead of having an `else` branch you could return early from this function

fho commented on internal/command/diff_inputs.go:
```diff
@@ -0,0 +1,387 @@
+}
+
+func getTaskRun(repo *baur.Repository, argDetails *diffInputArgDetails) (*storage.TaskRunWithID, error) {
+	if argDetails.task != nil {
```
instead of checking if task is nil here, only call the function when it should do something

StephenDiligent commented on internal/command/diff_inputs.go:
```diff
@@ -0,0 +1,387 @@
+			return fmt.Errorf("accepts 2 args, received %d", len(args))
+		}
+
+		containsWildCardPattern := "^\\*\\..+$|^.+\\.\\*$"
+		isNumericPattern := "^\\d+$"
+		containsTaskPattern := "^.+\\..+$"
+
+		for _, arg := range args {
+			containsWildCard, err := regexp.MatchString(containsWildCardPattern, arg)
+
+			if err != nil {
+				return err
+			}
+
+			if containsWildCard {
+				return fmt.Errorf("%s contains a wild card character, wild card characters are not allowed", arg)
+			}
+
+			isValid, err := regexp.MatchString(fmt.Sprintf("%s|%s", isNumericPattern, containsTaskPattern), arg)
+
+			if err != nil {
+				return err
+			}
+
+			if !isValid {
+				return fmt.Errorf("%s does not specify a task or task-run ID", arg)
+			}
+		}
```
If you are happy with the reduced error output and the fact that `app_one.build*` results in a match and will not cause an error but `app_one.*build` does not result in a match and returns an error then we can use your implementation.
The current implementation matches the asterisk only on either side of the period.

fho replied:
> If you are happy with the reduced error output

I think the reduced output is fine.

>  and the fact that app_one.build* results in a match and will not cause an error but app_one.*build does not result in a match and returns an error then we can use your implementation.

good catch, an end of string marker is needed after the  `<APP>.<TASK>` regex path and a beginning of string marker before the run-id part:
`^(\w+\.\w+)$|^[0-9]+\d*$`
https://regex101.com/r/D1DRNC/3



>  The current implementation matches the asterisk only on either side of the period.

What is reason for allowing to have an asterik?
With your implementation a call like `diff inputs -q random.build random.buil*` is allowed as valid argument.
But this can not resolve to the existing `random.build` task because this kind of syntax is not supported in the loader as wildcard.

StephenDiligent replied:
> With your implementation a call like `diff inputs -q random.build random.buil*` is allowed as valid argument.
> But this can not resolve to the existing `random.build` task because this kind of syntax is not supported in the loader as wildcard.

I was not aware that the app and task names cannot contain the asterisk character at all, they would have been treated literally instead of as a wildcard. I had taken from [here](https://github.com/simplesurance/baur/blob/master/internal/command/ls_runs.go#L192) that an asterisk by itself was not valid.
I'm more than happy for the asterisk character to be invalid in all cases, in which case your latest regex pattern looks good.

fho replied:
`ls apps` currently supports a syntax like:
- `*.build` -> to get the status of all build tasks for all apps
- `calc.*` -> to get the status of all tasks of the calc app

It does not work to use `*` as a wildcard for parts of the app or task names, like `calc.b*` will match `calc.build` and `calc.bla`.

For the `diff inputs` command one parameter should always resolve to a single task.
Supporting the wildcard syntax same way then for "ls apps" does not add much value here.
We would then also need to handle the case that the loader returned multiple tasks for a single parameter and print an error.

StephenDiligent replied:
I'll simplify things now and make the change.
I have tweaked your latest regex a little so it will only match on the whole argument and allow carets after the task
https://regex101.com/r/2VdBoe/1

StephenDiligent commented:
@fho I believe all of your comments have now been addressed.
However, it doesn't look like the checks are running

fho commented:
@StephenDiligent thanks a lot for the PR! :-)
